src/readcalls.c

```diff
@@ -81,322 +80,128 @@ void init_scoring(void) {
-    refreshp();
+    if (interactive) {
+	showstring("Reading logfile:", (char *)logfile);
+    }
 
     init_scoring();
 
     if ((fp = fopen(logfile, "r")) == NULL) {
 	showmsg("Error opening logfile ");
-	refreshp();
 	sleep(2);
 	exit(1);
     }
 
-    while (fgets(inputbuffer, LOGLINELEN + 1, fp) != NULL) {
+    bool log_changed = false;
+    // array of qso's
+    // FIXME use this instead of qsos[] (make it global and don't free it here)
+    // FIXME also include comments/notes into qso_t
+    GPtrArray *qso_array = g_ptr_array_new_with_free_func(qso_free);
+
+    while (fgets(inputbuffer, sizeof(inputbuffer), fp) != NULL) {
 
 	// drop trailing newline
 	inputbuffer[LOGLINELEN - 1] = '\0';
 
 	// remember logline in qsos[] field
 	g_strlcpy(qsos[linenr], inputbuffer, sizeof(qsos[0]));
 	linenr++;
+	// FIXME check for overflow....
 
-	show_progress(linenr);
+	if (interactive) {
+	    show_progress(linenr);
+	}
 
 	if (log_is_comment(inputbuffer))
-	    continue;		/* skip note in  log  */
-
-	// prepare helper variables
-	pfxnumcntidx = -1;
-	pxnr = 0;
+	    continue;		/* skip note in log */
 
 	qso = parse_qso(inputbuffer);
-	bandindex = qso->bandindex;
 
 	/* get the country number, not known at this point */
 	countrynr = getctydata(qso->call);
-
-	/*  lookup worked stations, add if new */
-	int station = lookup_or_add_worked(qso->call);
-	update_worked(station, qso);
-
-
-	if (continentlist_only) {
-	    if (!is_in_continentlist(dxcc_by_index(countrynr)->continent)) {
-		qsos_per_band[bandindex]++;
-		free_qso(qso);
-		qso = NULL;
-		continue;
-	    }
-	}
-
-	if (iscontest) {
-	    // get points
-	    total = total + log_get_points(inputbuffer);
-
-	    if (CONTEST_IS(CQWW) || itumult || wazmult) {
-		// get the zone
-		z = zone_nr(inputbuffer + 54);
-	    }
-
-	    if (wysiwyg_once || wysiwyg_multi ||
-		    unique_call_multi != 0 ||
-		    CONTEST_IS(ARRL_SS) ||
-		    serial_section_mult ||
-		    serial_grid4_mult ||
-		    sectn_mult ||
-		    sectn_mult_once ||
-		    (dx_arrlsections
-		     && ((countrynr == w_cty) || (countrynr == ve_cty)))) {
-		// get multi info
-		char *multbuffer = get_multi_from_line(inputbuffer);
-		remember_multi(multbuffer, bandindex, 0);
-		g_free(multbuffer);
-	    }
+	checkexchange(qso->comment, false);
+	if (strlen(normalized_comment) > 0) {   //FIXME global
+	    strcpy(qso->comment, normalized_comment);
 	}
-
-
-	if (pfxmultab) {
-	    getpx(qso->call);
-	    add_pfx(wpx_prefix, bandindex);
-	}
-
-
-	/* look if calls are excluded */
-	add_ok = true;
-
-	if (CONTEST_IS(ARRLDX_USA)
-		&& ((countrynr == w_cty) || (countrynr == ve_cty)))
-	    add_ok = false;
-
-	if (CONTEST_IS(PACC_PA)) {
-
-	    strcpy(hiscall, qso->call);
-
-	    add_ok = pacc_pa();
-
-	    if (add_ok == false) {
-		qsos_per_band[bandindex]++;
-	    }
-
-	    hiscall[0] = '\0';
+	dupe = is_dupe(qso->call, qso->bandindex, qso->mode);
+	addcall(qso);
+	score_qso();    //FIXME argument?
+
+	char *logline = makelogline(qso);
```

Comment:
After finishing all migration away from qsos[], we do not need to record such things as points, multis or prefixes in the written log. 
The we could drop these check for a change of logfile. The file would be the same as long as we do not change some of the qso information (time, call, exchange, band...).
But that leads us again to a discussion of logfile format. We should keep that for later.